### ISSUE TITLE
[Suggestion] Code Linter

### ISSUE THREAD

**Issue by armsnyder** [gitlabhq/terraform-provider-gitlab#361]
@ringods @roidelapluie 

How do the maintainers feel about possibly adding a code linter to the GitHub Actions? I don't think it needs to be super stringent, but enough to catch common mistakes. If you're open to it I'm happy to work on a PR that fixes current issues and another to add the new action.

Here's a preview of current issues that the [golangci-lint](https://github.com/golangci/golangci-lint) meta-linter reported with default configuration:

```
$ golangci-lint run
gitlab/resource_gitlab_user_test.go:158:6: `testAccCheckGitlabUserDestroy` is unused (deadcode)
func testAccCheckGitlabUserDestroy(s *terraform.State) error {
     ^
gitlab/resource_gitlab_deploy_key_enable.go:55:10: ineffectual assignment to `err` (ineffassign)
        key_id, err := strconv.Atoi(d.Get("key_id").(string))
                ^
gitlab/resource_gitlab_deploy_key_test.go:155:16: ineffectual assignment to `err` (ineffassign)
                deployKeyID, err := strconv.Atoi(rs.Primary.ID)
                             ^
gitlab/resource_gitlab_group_membership_test.go:111:11: ineffectual assignment to `err` (ineffassign)
                userId, err := strconv.Atoi(userIdString)
                        ^
gitlab/resource_gitlab_group_ldap_link.go:126:11: S1028: should use fmt.Errorf(...) instead of errors.New(fmt.Sprintf(...)) (gosimple)
                        return errors.New(fmt.Sprintf("LdapLink %s does not exist.", d.Id()))
                               ^
gitlab/resource_gitlab_pipeline_schedule.go:103:10: S1028: should use fmt.Errorf(...) instead of errors.New(fmt.Sprintf(...)) (gosimple)
                return errors.New(fmt.Sprintf("PipelineSchedule %d no longer exists in gitlab", pipelineScheduleID))
                       ^
gitlab/resource_gitlab_group_ldap_link_test.go:189:11: S1028: should use fmt.Errorf(...) instead of errors.New(fmt.Sprintf(...)) (gosimple)
                        return errors.New(fmt.Sprintf("LdapLink %s does not exist.", desiredLdapLinkId))
                               ^
gitlab/resource_gitlab_group_ldap_link_test.go:40:20: S1039: unnecessary use of fmt.Sprintf (gosimple)
                                                accessLevel: fmt.Sprintf("developer"),
                                                             ^
gitlab/resource_gitlab_group_ldap_link_test.go:51:20: S1039: unnecessary use of fmt.Sprintf (gosimple)
                                                accessLevel: fmt.Sprintf("maintainer"),
                                                             ^
gitlab/resource_gitlab_group_ldap_link_test.go:62:20: S1039: unnecessary use of fmt.Sprintf (gosimple)
                                                accessLevel: fmt.Sprintf("developer"),
                                                             ^
gitlab/resource_gitlab_group_ldap_link.go:97:10: S1034: assigning the result of this type assertion to a variable (switch err := err.(type)) could eliminate type assertions in switch cases (gosimple)
                switch err.(type) {
                       ^
gitlab/resource_gitlab_group_ldap_link.go:142:10: S1034: assigning the result of this type assertion to a variable (switch err := err.(type)) could eliminate type assertions in switch cases (gosimple)
                switch err.(type) {
                       ^
gitlab/helper_test.go:17:9: S1034: assigning the result of this type assertion to a variable (switch e := e.(type)) could eliminate type assertions in switch cases (gosimple)
        switch e.(type) {
               ^
gitlab/resource_gitlab_group_ldap_link_test.go:163:10: S1034: assigning the result of this type assertion to a variable (switch err := err.(type)) could eliminate type assertions in switch cases (gosimple)
                switch err.(type) {
                       ^
gitlab/data_source_gitlab_projects.go:102:2: S1031: unnecessary nil check around range (gosimple)
        if projects != nil {
        ^
gitlab/data_source_gitlab_group_test.go:13:13: S1025: the argument is already a string, there's no need to use fmt.Sprintf (gosimple)
        rString := fmt.Sprintf("%s", acctest.RandString(5))
                   ^
gitlab/data_source_gitlab_projects_test.go:138:82: S1025: the argument is already a string, there's no need to use fmt.Sprintf (gosimple)
                        errorMessageExpected.WriteString(fmt.Sprintf("%s=%v, ", attr, projectResource[fmt.Sprintf("%s", attr)]))
                                                                                                      ^
gitlab/data_source_gitlab_user_test.go:13:13: S1025: the argument is already a string, there's no need to use fmt.Sprintf (gosimple)
        rString := fmt.Sprintf("%s", acctest.RandString(5))
                   ^
gitlab/resource_gitlab_deploy_token.go:48:19: SA1019: validation.ValidateRFC3339TimeString is deprecated: use IsRFC3339Time() instead  (staticcheck)
                                ValidateFunc: validation.ValidateRFC3339TimeString,
                                              ^
gitlab/resource_gitlab_project.go:321:3: SA1019: d.SetPartial is deprecated: Partial state has very limited benefit given Terraform refreshes before operations by default.  (staticcheck)
                d.SetPartial(setProperty)
                ^
gitlab/resource_gitlab_project.go:334:3: SA1019: d.SetPartial is deprecated: Partial state has very limited benefit given Terraform refreshes before operations by default.  (staticcheck)
                d.SetPartial("shared_with_groups")
                ^
gitlab/resource_gitlab_project.go:345:3: SA1019: d.SetPartial is deprecated: Partial state has very limited benefit given Terraform refreshes before operations by default.  (staticcheck)
                d.SetPartial(("archived"))
                ^
gitlab/resource_gitlab_group_variable_test.go:125:3: SA9003: empty branch (staticcheck)
                if err == nil {
                ^
```

**Comment by willianpaixao** [comment]
Fix #642

**Comment by ringods** [comment]
@armsnyder FYI this provider has been officially transferred to the Gitlab team. I am no longer a maintainer of it.
